core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,404 @@
+      this.expireAfterAccessMs = expireAfterAccessMs;
+      this.maxTotalBytes = maxTotalBytes;
+      this.maxContentLength = maxContentLength;
+
+      Caffeine<Object, Object> builder = Caffeine.newBuilder();
+      if (expireAfterAccessMs > 0) {
+        builder = builder.expireAfterAccess(Duration.ofMillis(expireAfterAccessMs));
+      }
+
+      this.cache = builder.maximumWeight(maxTotalBytes)
+          .weigher((Weigher<String, byte[]>) (key, value) -> value.length)
+          .recordStats()
+          .build();
+    }
+
+    public long expireAfterAccess() {
+      return expireAfterAccessMs;
+    }
+
+    public long maxContentLength() {
+      return maxContentLength;
+    }
+
+    public long maxTotalBytes() {
+      return maxTotalBytes;
+    }
+
+    public Cache<String, byte[]> cache() {
+      return cache;
+    }
+
+    public CacheStats stats() {
+      return cache.stats();
+    }
+
+    @Override
+    public String toString() {
+      return getClass().getSimpleName() + '{' +
+          "expireAfterAccessMs=" + expireAfterAccessMs + ", " +
+          "maxContentLength=" + maxContentLength + ", " +
+          "maxTotalBytes=" + maxTotalBytes + ", " +
+          "cacheStats=" + cache.stats() + '}';
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length : wrappedInputFile.getLength();
+    }
+
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        return wrappedInputFile.newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      }
+    }
+
+    @Override
+    public String location() {
+      return wrappedInputFile.location();
+    }
+
+    @Override
+    public boolean exists() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return buf != null || wrappedInputFile.exists();
+    }
+
+    public long getLengthChecked() throws IOException {
+      try {
+        return getLength();
+      } catch (RuntimeIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    private byte[] getFileContent() {
+      try {
+        int totalByteRead = 0;
+        boolean eos = false;
+        int fileLength = (int) getLength();
+        byte[] buf = new byte[fileLength];
```

Comment:
I would prefer not to use a single byte[] allocation. If `getLength` is incorrect (which can evidently be the case for some encrypted streams) then we need to be able to handle that case. Plus, this could easily result in humongous allocations in the JVM that are hard to garbage collect. It's better to allocate ByteBuffer instances as needed and use a stream like Parquet's `ByteBufferInputStream` to stitch them together.

Reply:
Yeah, thinking about this more, I think it should allocate small buffers. In a loop, allocate a buffer that is 4 or 8 MB, then call `ByteStreams.readFully` to fill the buffer, and then wrap it in a `ByteBuffer`. Then use `ByteBufferInputStream` and the associated classes from Parquet (probably just copy those into `util` along with tests since we will use them in other places).

You may need a cache entry to have a fast size lookup for the caffeine `Cache`.

Reply:
Are you referring to this class?
https://github.com/apache/parquet-mr/blob/master/parquet-common/src/main/java/org/apache/parquet/bytes/ByteBufferInputStream.java

I look around, but can't seem to find class that can adapt that `ByteBufferInputStream` to a `SeekableInputStream` needed for `newStream()` later. Can you confirm that I should copy that class (possibly with `SingleBufferInputStream` and `MultiBufferInputStream` as well), and implement `seek()` on it so it can be adapted to `SeekableInputStream`?

Reply:
Refactored to use `IOUtil.readRemaining`. However, it still relies on `getLength` being accurate. Please let me know if it should keep looping until EOF met.